File: c++/2_ds/2_level_order_traversal.cpp
// Theory:
// A binary tree is a hierarchical structure where each node has at most
// two children. Traversing it "level order" (breadth-first, top to
// bottom, left to right within each level) needs a queue, not a stack:
// you process a node, then enqueue its children so they're visited after
// every other node already waiting at the current depth - a FIFO
// structure preserves that "finish this level before starting the next"
// ordering. This is different from depth-first traversals (preorder,
// inorder, postorder), which use a stack (explicit or the call stack) and
// dive to a leaf before backtracking. BFS-style level order is the
// natural choice whenever "shortest path" or "distance/depth" matters,
// since it explores everything at distance k before anything at k+1.

// Headers & STL components:
// <queue> - std::queue, O(1) push/pop, the queue that drives BFS
// <optional> - std::optional<int>, represents a possibly-missing tree value when building test trees
// <vector> - the input/output container type here

// Task:
// Given the root of a binary tree (TreeNode with val/left/right), return
// its level order traversal as a vector of vectors - one inner vector per
// depth, left to right.
// Test1: Inputs: root = {3,9,20,-,-,15,7}   Outputs: {{3},{9,20},{15,7}}
// Test2: Inputs: root = {1}   Outputs: {{1}}
// Test3: Inputs: root = {}   Outputs: {}
// Hint:
// Ugly way: recursive DFS - track each node's depth, append its value
// into a vector indexed/grouped by depth.
// Right way: std::queue - BFS, processing one full level (queue
// snapshot) per iteration.

// Solutions:
#include <cassert>
#include <iostream>
#include <optional>
#include <queue>
#include <vector>

struct TreeNode {
    int val;
    TreeNode* left;
    TreeNode* right;
    explicit TreeNode(int v) : val(v), left(nullptr), right(nullptr) {}
};

TreeNode* build_tree(const std::vector<std::optional<int>>& values) {
    if (values.empty() || !values[0].has_value()) return nullptr;
    auto* root = new TreeNode(*values[0]);
    std::queue<TreeNode*> q;
    q.push(root);
    size_t i = 1;
    while (!q.empty() && i < values.size()) {
        TreeNode* node = q.front();
        q.pop();
        if (i < values.size() && values[i].has_value()) {
            node->left = new TreeNode(*values[i]);
            q.push(node->left);
        }
        ++i;
        if (i < values.size() && values[i].has_value()) {
            node->right = new TreeNode(*values[i]);
            q.push(node->right);
        }
        ++i;
    }
    return root;
}

// Space - Time Complexity analysis:
// space:
// time:

std::vector<std::vector<int>> level_order_ugly([[maybe_unused]] TreeNode* root) {
    // TODO: implement
    return {};
}

// Space - Time Complexity analysis:
// space:
// time:

std::vector<std::vector<int>> level_order(TreeNode* root) {
    return level_order_ugly(root);
}

// Tests:
void validate() {
    struct Case {
        std::vector<std::optional<int>> values;
        std::vector<std::vector<int>> expected;
    };
    std::vector<Case> cases = {
        {{3, 9, 20, std::nullopt, std::nullopt, 15, 7}, {{3}, {9, 20}, {15, 7}}},
        {{1}, {{1}}},
        {{}, {}},
        {{1, 2, 3}, {{1}, {2, 3}}},
    };
    for (const auto& c : cases) {
        assert(level_order_ugly(build_tree(c.values)) == c.expected);
        assert(level_order(build_tree(c.values)) == c.expected);
    }
    std::cout << "SUCCESS" << std::endl;
}

int main() {
    validate();
    return 0;
}

File: python/ds/2_level_order_traversal.py
# Theory:
## A binary tree is a hierarchical structure where each node has at most
## two children. Traversing it "level order" (breadth-first, top to
## bottom, left to right within each level) needs a queue, not a stack:
## you process a node, then enqueue its children so they're visited after
## every other node already waiting at the current depth - a FIFO
## structure preserves that "finish this level before starting the next"
## ordering. This is different from depth-first traversals (preorder,
## inorder, postorder), which use a stack (explicit or the call stack) and
## dive to a leaf before backtracking. BFS-style level order is the
## natural choice whenever "shortest path" or "distance/depth" matters,
## since it explores everything at distance k before anything at k+1.


# Packages:
## collections.deque - O(1) append/popleft, the queue that drives BFS
## collections.defaultdict(list) - grouping nodes by level when building the result
## queue.Queue - the thread-safe alternative to deque (unneeded here, single-threaded)


# Task:
## Given the root of a binary tree (TreeNode with val/left/right), return
## its level order traversal as a list of lists - one inner list per
## depth, left to right.
## Test1: Inputs: root = [3,9,20,None,None,15,7]   Outputs: [[3],[9,20],[15,7]]
## Test2: Inputs: root = [1]   Outputs: [[1]]
## Test3: Inputs: root = []   Outputs: []
## Hint:
## Ugly way: recursive DFS - track each node's depth, append its value
## into a list indexed/grouped by depth.
## Right way: collections.deque - BFS, processing one full level (queue
## snapshot) per iteration.


class TreeNode:
    def __init__(self, val=0, left=None, right=None):
        self.val = val
        self.left = left
        self.right = right


def build_tree(values):
    # level-order list with None for missing children, LeetCode-style
    if not values or values[0] is None:
        return None
    root = TreeNode(values[0])
    queue = [root]
    i = 1
    while queue and i < len(values):
        node = queue.pop(0)
        if i < len(values) and values[i] is not None:
            node.left = TreeNode(values[i])
            queue.append(node.left)
        i += 1
        if i < len(values) and values[i] is not None:
            node.right = TreeNode(values[i])
            queue.append(node.right)
        i += 1
    return root


# Solutions:

# Space - Time Complexity analysis:
## space:
## time:

def level_order_ugly(root):
    pass


# Space - Time Complexity analysis:
## space:
## time:

def level_order(root):
    return level_order_ugly(root)


# Tests:
def validate():
    cases = [
        ([3, 9, 20, None, None, 15, 7], [[3], [9, 20], [15, 7]]),
        ([1], [[1]]),
        ([], []),
        ([1, 2, 3], [[1], [2, 3]]),
    ]
    for fn in (level_order_ugly, level_order):
        for values, expected in cases:
            result = fn(build_tree(values))
            assert result == expected, (
                f"{fn.__name__}({values!r}): expected {expected}, got {result}"
            )
    print("SUCCESS")


if __name__ == "__main__":
    # validate()
    print(level_order_ugly(build_tree([3, 9, 20, None, None, 15, 7])))
    print(level_order_ugly(build_tree([1])))
    print(level_order_ugly(build_tree([])))
    print(level_order_ugly(build_tree([1, 2, 3])))
